clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -1573,5 +1584,163 @@ invariant_checker_t::check_for_pc_discontinuity(
+                uint eflags = instr_get_arith_flags(noalloc_instr, DR_QUERY_DEFAULT);
+                report_if_false(
+                    shard,
+                    eflags == 0 || eflags == EFLAGS_WRITE_ARITH ||
+                        eflags == EFLAGS_READ_ARITH ||
+                        eflags == (EFLAGS_WRITE_ARITH | EFLAGS_READ_ARITH),
+                    "DR_ISA_REGDEPS instruction has incorrect arithmetic flags");
+                // Instruction length should be a multiple of 4 bytes
+                // (i.e., REGDEPS_ALIGN_BYTES).
+                report_if_false(
+                    shard, ((next_pc - encoding_addr) % REGDEPS_ALIGN_BYTES) == 0,
+                    "DR_ISA_REGDEPS instruction has incorrect length, it's not a "
+                    "multiple of REGDEPS_ALIGN_BYTES = 4");
+            }
+        }
+    }
+
+    // Check markers.
+    if (memref.marker.type == TRACE_TYPE_MARKER) {
+        switch (memref.marker.marker_type) {
+        case TRACE_MARKER_TYPE_SYSCALL_IDX:
+            report_if_false(shard, false,
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                            "TRACE_MARKER_TYPE_SYSCALL_IDX markers");
+            break;
+        case TRACE_MARKER_TYPE_SYSCALL:
+            report_if_false(shard, false,
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                            "TRACE_MARKER_TYPE_SYSCALL markers");
+            break;
+        case TRACE_MARKER_TYPE_SYSCALL_TRACE_START:
+            report_if_false(shard, false,
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                            "TRACE_MARKER_TYPE_SYSCALL_TRACE_START markers");
+            break;
+        case TRACE_MARKER_TYPE_SYSCALL_TRACE_END:
+            report_if_false(shard, false,
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                            "TRACE_MARKER_TYPE_SYSCALL_TRACE_END markers");
+            break;
+        case TRACE_MARKER_TYPE_SYSCALL_FAILED:
+            report_if_false(shard, false,
+                            "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                            "TRACE_MARKER_TYPE_SYSCALL_FAILED markers");
+            break;
+        case TRACE_MARKER_TYPE_FUNC_ID:
+#ifdef LINUX
+            report_if_false(
+                shard,
+                memref.marker.marker_value ==
+                    static_cast<uintptr_t>(func_trace_t::TRACE_FUNC_ID_SYSCALL_BASE) +
+                        SYS_futex,
+                "OFFLINE_FILE_TYPE_ARCH_REGDEPS traces cannot have "
+                "TRACE_MARKER_TYPE_FUNC_ID markers related to functions that "
+                "are not SYS_futex");
+#else
+            report_if_false(shard, false,
+                            "The function ID cannot be determined for the "
+                            "TRACE_MARKER_TYPE_FUNC_ID marker");
```

Comment:
This will fire every time for non-LINUX, which is what we want to do with the message at line 1681: but the message here is incorrect.  Best to share the string at line 1681 and have the bool passed to report_if_false be set based on the arch.

Reply:
Rationale for having 2 different messages: https://github.com/DynamoRIO/dynamorio/pull/6889#discussion_r1682041354